pkg/kv/kvclient/kvcoord: BenchmarkTxnWriteBuffer failed

pkg/kv/kvclient/kvcoord.BenchmarkTxnWriteBuffer [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachprodMicrobenchmarkWeeklyWip/20958673?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachprodMicrobenchmarkWeeklyWip/20958673?buildTab=artifacts#./artifacts/microbench/experiment) on master @ [94a1e22436ad935d8c7f9524cb8fc0fd7b24faaa](https://github.com/cockroachdb/cockroach/commits/94a1e22436ad935d8c7f9524cb8fc0fd7b24faaa):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x29b9e57]

```
Stack: 

```
goroutine 143 gp=0xc000a032c0 m=29 mp=0xc00145e008 [running]:
panic({0x55fcfe0?, 0xa76beb0?})
	GOROOT/src/runtime/panic.go:802 +0x168 fp=0xc0011f55d8 sp=0xc0011f5528 pc=0x4baba8
runtime.panicmem(...)
	GOROOT/src/runtime/panic.go:262
runtime.sigpanic()
	GOROOT/src/runtime/signal_unix.go:925 +0x359 fp=0xc0011f5638 sp=0xc0011f55d8 pc=0x4bd779
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.requireAllFlushedRequestsProcessed({0xc001829208?, 0x2?, 0x2?})
	pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go:1873 +0x57 fp=0xc0011f56e8 sp=0xc0011f5638 pc=0x29b9e57
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnWriteBuffer).flushBufferAndSendBatch(0xc0011f5ca0, {0x6c7b6b0, 0xae74940}, 0xc001a3eb40)
	pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go:1862 +0xcb2 fp=0xc0011f59f0 sp=0xc0011f56e8 pc=0x29b9b12
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.BenchmarkTxnWriteBuffer.func6(0xc0016ec2c8)
	pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go:2902 +0x348 fp=0xc0011f5f10 sp=0xc0011f59f0 pc=0x2a9a5c8
testing.(*B).runN(0xc0016ec2c8, 0x1)
	GOROOT/src/testing/benchmark.go:219 +0x190 fp=0xc0011f5fa0 sp=0xc0011f5f10 pc=0x5be1f0
testing.(*B).run1.func1()
	GOROOT/src/testing/benchmark.go:245 +0x48 fp=0xc0011f5fe0 sp=0xc0011f5fa0 pc=0x5be868
runtime.goexit({})
	src/runtime/asm_amd64.s:1693 +0x1 fp=0xc0011f5fe8 sp=0xc0011f5fe0 pc=0x4c4141
created by testing.(*B).run1 in goroutine 57
	GOROOT/src/testing/benchmark.go:238 +0x90
```
<details><summary>Log preceding fatal error</summary>
<p>

```
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=0.00/from_buffer=50.00/from_batch=0.00-32        	   19900	     60002 ns/op	   72016 B/op	      27 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=0.00/from_buffer=50.00/from_batch=50.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=0.00/from_buffer=50.00/from_batch=50.00-32       	   19897	     59199 ns/op	   72016 B/op	      27 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=0.00/from_buffer=50.00/from_batch=100.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=0.00/from_buffer=50.00/from_batch=100.00-32      	   19824	     59587 ns/op	   72016 B/op	      27 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=0.00/from_buffer=100.00/from_batch=0.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=0.00/from_buffer=100.00/from_batch=0.00-32       	   19828	     60021 ns/op	   72016 B/op	      27 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=0.00/from_buffer=100.00/from_batch=50.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=0.00/from_buffer=100.00/from_batch=50.00-32      	   19933	     59819 ns/op	   72016 B/op	      27 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=0.00/from_buffer=100.00/from_batch=100.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=0.00/from_buffer=100.00/from_batch=100.00-32     	   19806	     60446 ns/op	   72016 B/op	      27 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=0.00/from_batch=0.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=0.00/from_batch=0.00-32        	   15926	     75031 ns/op	   90800 B/op	     245 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=0.00/from_batch=50.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=0.00/from_batch=50.00-32       	   16026	     75243 ns/op	   90800 B/op	     245 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=0.00/from_batch=100.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=0.00/from_batch=100.00-32      	   15842	     75343 ns/op	   90800 B/op	     245 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=50.00/from_batch=0.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=50.00/from_batch=0.00-32       	   10000	    107499 ns/op	  103691 B/op	     407 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=50.00/from_batch=50.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=50.00/from_batch=50.00-32      	   10000	    105662 ns/op	  102187 B/op	     424 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=50.00/from_batch=100.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=50.00/from_batch=100.00-32     	   12274	     97080 ns/op	  102258 B/op	     441 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=100.00/from_batch=0.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=100.00/from_batch=0.00-32      	    8703	    126495 ns/op	  117859 B/op	     561 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=100.00/from_batch=50.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=100.00/from_batch=50.00-32     	    8322	    133908 ns/op	  115610 B/op	     594 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=100.00/from_batch=100.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=50.00/from_buffer=100.00/from_batch=100.00-32    	    9304	    120249 ns/op	  119427 B/op	     630 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=0.00/from_batch=0.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=0.00/from_batch=0.00-32       	   13610	     88514 ns/op	  109760 B/op	     447 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=0.00/from_batch=50.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=0.00/from_batch=50.00-32      	   13342	     88914 ns/op	  109760 B/op	     447 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=0.00/from_batch=100.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=0.00/from_batch=100.00-32     	   13653	     88242 ns/op	  109760 B/op	     447 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=50.00/from_batch=0.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=50.00/from_batch=0.00-32      	    7003	    157676 ns/op	  136826 B/op	     763 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=50.00/from_batch=50.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=50.00/from_batch=50.00-32     	    6442	    169726 ns/op	  133296 B/op	     796 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=50.00/from_batch=100.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=50.00/from_batch=100.00-32    	    7393	    140443 ns/op	  132632 B/op	     830 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=100.00/from_batch=0.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=100.00/from_batch=0.00-32     	    5322	    202635 ns/op	  163236 B/op	    1067 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=100.00/from_batch=50.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=100.00/from_batch=50.00-32    	    5020	    214924 ns/op	  158940 B/op	    1141 allocs/op
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=100.00/from_batch=100.00
BenchmarkTxnWriteBuffer/SendLocked/size=8/reads=100.00/from_buffer=100.00/from_batch=100.00-32   	    5658	    187726 ns/op	  165116 B/op	    1210 allocs/op
BenchmarkTxnWriteBuffer/flushBufferAndSendBatch/size=8

I260111 16:49:05.214733 1 (gostd) rand.go:288  [-] 1  random seed: 4962876601786616905
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*BenchmarkTxnWriteBuffer.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58597

# Investigation Summary - Issue #160881

**Date:** 2026-01-12  
**Test:** `pkg/kv/kvclient/kvcoord.BenchmarkTxnWriteBuffer`  
**Type:** Microbenchmark (unit test)  
**SHA:** `94a1e22436ad935d8c7f9524cb8fc0fd7b24faaa`  
**Priority:** Release-blocker

## Key Findings

The benchmark's mock sender returns **zero-valued `ResponseUnion`** for non-`GetRequest` requests (like `PutRequest`), causing a nil pointer dereference when the `requireAllFlushedRequestsProcessed` assertion calls `resp.GetInner()`.

### Timeline

1. **May 6, 2025** (`a132ee17c59`): Benchmark added by @miraradeva with a mock sender that only handles `GetRequest` responses
2. **Aug 15, 2025** (`c6ba3771cc5`): `requireAllFlushedRequestsProcessed` assertion added by @stevendanna - calls `resp.GetInner().Header().ResumeSpan` on all responses
3. **Jan 11, 2026** (`94a1e22436a`): Panic occurs when `GetInner()` returns `nil` for non-Get responses

### The Bug (in test code, not production)

In `txn_interceptor_write_buffer_test.go` lines 2768-2788:

```go
sendFunc := func(ba *kvpb.BatchRequest) (*kvpb.BatchResponse, *kvpb.Error) {
    br := ba.CreateReply()
    var resps []kvpb.ResponseUnion
    resp := kvpb.ResponseUnion{}  // ← Single resp, REUSED
    for _, req := range ba.Requests {
        switch req.GetInner().(type) {
        case *kvpb.GetRequest:
            // Only GetRequest sets resp.Value!
            resp.Value = &kvpb.ResponseUnion_Get{Get: &kvpb.GetResponse{...}}
        }
        // For non-Get requests, resp remains zero-valued
        resps = append(resps, resp)  // ← Zero-valued for Put/Delete!
    }
    br.Responses = resps
    return br, nil
}
```

The `flushBufferAndSendBatch` subtest (lines 2888-2909) flushes buffered **writes** (PutRequests), but the mock returns zero-valued `ResponseUnion` for them.

In `txn_interceptor_write_buffer.go` line 1873:
```go
func requireAllFlushedRequestsProcessed(responses []kvpb.ResponseUnion) error {
    for _, resp := range responses {
        if resp.GetInner().Header().ResumeSpan != nil {  // ← PANIC: GetInner() returns nil!
```

## Related Issues/PRs

- **#151650** - Prior `BenchmarkTxnWriteBuffer` failures (timeouts, not panics) - **CLOSED**
- **PR #160648** - Fixed timeout issue by removing timer start/stop - **MERGED** Jan 9, 2026
- The current panic (#160881) is a **different failure** from #151650

## Root Cause Analysis

**Not a regression in production code** - this is a **test infrastructure bug**.

The benchmark's mock sender was written before the `requireAllFlushedRequestsProcessed` assertion was added. When the assertion was added, the benchmark wasn't updated to return proper responses for all request types.

### Why did this only surface now?

The `flushBufferAndSendBatch` subtest was likely not running frequently enough (or was timing out before reaching the assertion) until PR #160648 fixed the timeout issue on Jan 9, 2026. Once the benchmark started completing, it hit this assertion failure.

## Evidence

**Stack trace:**
```
panic: runtime error: invalid memory address or nil pointer dereference
...
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.requireAllFlushedRequestsProcessed
    pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go:1873
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord.(*txnWriteBuffer).flushBufferAndSendBatch
    pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go:1862
```

## Recommendations

### Fix

Update the mock sender in `BenchmarkTxnWriteBuffer` to return proper responses for **all** request types:

```go
sendFunc := func(ba *kvpb.BatchRequest) (*kvpb.BatchResponse, *kvpb.Error) {
    br := ba.CreateReply()
    br.Txn = ba.Txn
    var resps []kvpb.ResponseUnion
    for _, req := range ba.Requests {
        resp := kvpb.ResponseUnion{}  // ← Create new resp for each request
        switch r := req.GetInner().(type) {
        case *kvpb.GetRequest:
            v := makeValue(kvSize)
            resp.Value = &kvpb.ResponseUnion_Get{Get: &kvpb.GetResponse{Value: &v}}
        case *kvpb.PutRequest:
            resp.Value = &kvpb.ResponseUnion_Put{Put: &kvpb.PutResponse{}}
        case *kvpb.DeleteRequest:
            resp.Value = &kvpb.ResponseUnion_Delete{Delete: &kvpb.DeleteResponse{}}
        // Add other request types as needed
        default:
            // Use CreateReply() for proper response type
            resp = req.CreateReply()
        }
        resps = append(resps, resp)
    }
    br.Responses = resps
    return br, nil
}
```

**Alternative (simpler):** Use `ba.CreateReply()` which already creates properly typed responses:
```go
sendFunc := func(ba *kvpb.BatchRequest) (*kvpb.BatchResponse, *kvpb.Error) {
    br := ba.CreateReply()  // ← This creates proper responses for ALL request types!
    br.Txn = ba.Txn
    // Optionally populate Get responses with values if needed
    for i, req := range ba.Requests {
        if _, ok := req.GetInner().(*kvpb.GetRequest); ok {
            v := makeValue(kvSize)
            br.Responses[i].GetGet().Value = &v
        }
    }
    return br, nil
}
```

### Assignee

This should be assigned to **@miraradeva** (original benchmark author) or **@stevendanna** (assertion author).

### Severity

- **Test-only bug** - no production impact
- **Release-blocker** label should be removed after fix, as this is flaky test infrastructure, not a product bug

--- a/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go
+++ b/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer_test.go
@@ -2766,24 +2766,19 @@ func BenchmarkTxnWriteBuffer(b *testing.B) {
 	makeBuffer := func(kvSize int, txn *roachpb.Transaction, numWrites int) txnWriteBuffer {
 		twb, mockSender, _ := makeMockTxnWriteBuffer(ctx, metrics)
 		sendFunc := func(ba *kvpb.BatchRequest) (*kvpb.BatchResponse, *kvpb.Error) {
+			// Use CreateReply() which creates properly-typed responses for all
+			// request types. This is important because requireAllFlushedRequestsProcessed
+			// calls GetInner() on each response, which would panic on zero-valued
+			// ResponseUnion entries.
 			br := ba.CreateReply()
 			br.Txn = ba.Txn
-			var resps []kvpb.ResponseUnion
-			resp := kvpb.ResponseUnion{}
-			// All requests get responses. Gets also have a return value.
-			for _, req := range ba.Requests {
-				switch req.GetInner().(type) {
-				case *kvpb.GetRequest:
+			// Populate Get responses with values.
+			for i, req := range ba.Requests {
+				if _, ok := req.GetInner().(*kvpb.GetRequest); ok {
 					v := makeValue(kvSize)
-					resp.Value = &kvpb.ResponseUnion_Get{
-						Get: &kvpb.GetResponse{
-							Value: &v,
-						},
-					}
+					br.Responses[i].GetGet().Value = &v
 				}
-				resps = append(resps, resp)
 			}
-			br.Responses = resps
 			return br, nil
 		}
 		mockSender.MockSend(sendFunc)

